doc/developer-guides/coding_guidelines.rst

```diff
@@ -1121,6 +1428,312 @@ Compliant example::
+       uint32_t showcase_unsigned;
+       
+       showcase_unsigned = (uint32_t)(showcase_signed_1 - showcase_signed_2);
+
+
+TY-15: A complex unsigned integer expression shall not be cast to signed type
+=============================================================================
+
+Compliant example::
+
+    uint32_t showcase_unsigned_1 = 10U;
+    uint32_t showcase_unsigned_2 = 5U;
+    int32_t showcase_signed;
+    
+    showcase_signed = (int32_t)showcase_unsigned_1 - (int32_t)showcase_unsigned_2;
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       uint32_t showcase_unsigned_1 = 10U;
+       uint32_t showcase_unsigned_2 = 5U;
+       int32_t showcase_signed;
+       
+       showcase_signed = (int32_t)(showcase_unsigned_1 - showcase_unsigned_2);
+
+
+TY-16: A complex integer expression shall not be cast to types other than integer
+=================================================================================
+
+Compliant example::
+
+    /* 0x61 is 'a' in ASCII Table */
+    uint32_t showcase_u32;
+    char showcase_char;
+    
+    showcase_u32 = 0x61U + 1U;
+    showcase_char = (char)showcase_u32;
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       /* 0x61 is 'a' in ASCII Table */
+       uint32_t showcase_u32;
+       char showcase_char;
+       
+       showcase_u32 = 0x61U;
+       showcase_char = (char)(showcase_u32 + 1U);
+
+
+TY-17: Integer shall not be used when a plain char is expected
+==============================================================
+
+Compliant example::
+
+    char showcase;
+    
+    switch (showcase) {
+    case 'a':
+        /* do something */
+        break;
+    case 'A':
+        /* do something */
+        break;
+    default:
+        break;
+    }
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       char showcase;
+       
+       switch (showcase) {
+       /* 0x61 is 'a' in ASCII Table */
+       case 0x61:
+           /* do something */
+           break;
+       case 'A':
+           /* do something */
+           break;
+       default:
+           break;
+       }
+
+
+TY-18: A pointer shall not be cast to an integer
+================================================
+
+Compliant example::
+
+    uint64_t *showcase_ptr;
+    
+    uint64_t showcase = *showcase_ptr;
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       uint64_t *showcase_ptr;
+       
+       uint64_t showcase = (uint64_t)showcase_ptr;
+
+
+TY-19: An integer shall not be cast to a pointer
+================================================
+
+Compliant example::
+
+    uint64_t showcase = 10UL;
+    
+    uint64_t *showcase_ptr = &showcase;
+
+.. rst-class:: non-compliant-code
+
+   Non-compliant example::
+
+       uint64_t showcase = 10UL;
+       
+       uint64_t *showcase_ptr = (uint64_t *)showcase;
+
+
+TY-20: All types declared by typedef shall be used
+==================================================
+
```

Comment:
An explanation could be:
```
Typedefs that are not used shall be deleted.
```

Typedefs could be used in another file though, right? So how is this rule detected and enforced?

Reply:
I will add this explanation in next version.
Yes, typedef is usually defined in header file and can be used in other files. 
We are using the scanning tool to detect all the violations. In current CI system, every hypervisor PR is scanned by the tool and the violations are in the comments.